docs/reference/ingest/processors/reroute.asciidoc

```diff
@@ -0,0 +1,93 @@
+If the document contains a `event.dataset` value, it will be updated to reflect the same value as `data_stream.dataset`.
+
+Note that the client needs to have permissions to the final target.
+Otherwise, the document will be rejected with a security exception which looks like this:
+
+[source,js]
+--------------------------------------------------
+{"type":"security_exception","reason":"action [indices:admin/auto_create] is unauthorized for API key id [8-dt9H8BqGblnY2uSI--] of user [elastic/fleet-server] on indices [logs-foo-default], this action is granted by the index privileges [auto_configure,create_index,manage,all]"}
+--------------------------------------------------
+// NOTCONSOLE
+
+[[reroute-options]]
+.Reroute options
+[options="header"]
+|======
+| Name          | Required  | Default                      | Description
+| `destination` | no        | -                            | A static value for the target. Can't be set when the `dataset` or `namespace` option is set.
+| `dataset`     | no        | `{{data_stream.dataset}}`   a| Field references or a static value for the dataset part of the data stream name. In addition to the criteria for <<indices-create-api-path-params, index names>>, cannot contain `-` and must be no longer than 100 characters. Example values are `nginx.access` and `nginx.error`.
+
+Supports field references with a mustache-like syntax (denoted as `{{double}}` or `{{{triple}}}` curly braces). When resolving field references, the processor replaces invalid characters with `_`. Uses the `<dataset>` part of the index name as a fallback if all field references resolve to a `null`, missing, or non-string value.
+| `namespace`   | no        | `{{data_stream.namespace}}` a| Field references or a static value for the namespace part of the data stream name. See the criteria for <<indices-create-api-path-params, index names>> for allowed characters. Must be no longer than 100 characters.
+
+Supports field references with a mustache-like syntax (denoted as `{{double}}` or `{{{triple}}}` curly braces). When resolving field references, the processor replaces invalid characters with `_`. Uses the `<namespace>` part of the index name as a fallback if all field references resolve to a `null`, missing, or non-string value.
+include::common-options.asciidoc[]
+|======
+
+The `if` option can be used to define the condition in which the document should be rerouted to a new target.
+
+[source,js]
+--------------------------------------------------
+{
+  "reroute": {
+    "tag": "nginx",
+    "if" : "ctx?.log?.file?.path?.contains('nginx')",
+    "dataset": "nginx"
+  }
+}
+--------------------------------------------------
+// NOTCONSOLE
+
+The dataset and namespace options can contain either a single value or a list of values that are used as a fallback.
+If a field reference evaluates to `null`, is not present in the document, or if the value is not a `String`, the next value or field reference is used.
```

Comment:
I'm strongly in favor of not dropping data. The user setting up the rule is not in control of the data that is shipped. It can happen that most events are fine but there are some events which have malformed data like by accident an array. 

Ideally, there is an easy way for users to debug this or even see it in the docs that got ingested. That is where I would focus the discussion but we can still follow up on this one.

We can document it as a best practice to have a string without variables inside as last fallback option if the fallback option is not the target data stream.

Reply:
> We can document it as a best practice to have a string without variables inside as last fallback option if the fallback option is not the target data stream.

We can, but there's always the implicit fallback of getting the value from the current target (`_index: <type>-<dataset>-<namespace>`)

Reply:
I think trying to treat this with leniency is taking some of the data stream failure store functionality and putting it into this reroute processor. I would rather that we throw and exception and fail in that scenario. We can then handle this in a better way once we have that functionality.

Imagine N months from now when we do have a failure store, this leniency means that tracking the location of the data down to where it failed is now harder, because we silently skip a service name parameter with no way to make it "re-strict". If we are strict now, we could always decide at a later time to relax that strictness and allow more leniency, however, we can't go the other direction.

We already have this same sort of strictness in our documents by making things `constant_keyword` (for example, if someone were to write a doc where `data_stream.dataset` was an array).

Reply:
We've discussed this today and decided to change it to be more strict, as suggested by Joe.